src/WebSockets/Channels/PresenceChannel.php

```diff
@@ -26,14 +48,29 @@ public function subscribe(ConnectionInterface $connection, stdClass $payload)
         $channelData = json_decode($payload->channel_data);
         $this->users[$connection->socketId] = $channelData;
 
-        // Send the success event
-        $connection->send(json_encode([
-            'event' => 'pusher_internal:subscription_succeeded',
-            'channel' => $this->channelName,
-            'data' => json_encode($this->getChannelData()),
-        ]));
-
-        $this->broadcastToOthers($connection, [
+        // Add the connection as a member of the channel
+        $this->replication
+            ->joinChannel(
+                $connection->app->id,
+                $this->channelName,
+                $connection->socketId,
+                json_encode($channelData)
+            );
+
+        // We need to pull the channel data from the replication backend,
+        // otherwise we won't be sending the full details of the channel
+        $this->replication
+            ->channelMembers($connection->app->id, $this->channelName)
```

Comment:
In picking this up and trying to use presence channels, I observed a problem here which prevents the client from ever receiving the `subscription_succeeded` event.

The `channelMembers` implementation in `RedisClient` is executing an `hgetall` which returns keys and values in a single array. `$users` is then passed through here to `getChannelData` which fails since `getUserIds` is expecting a `user_id` property on every item. The array looks like this from `hgetall`:

```
[
  237207758.620121671,
  {"user_id":"1","user_info":{"id":"1","name":"Albert"}},
  38678372.334185943,
  {"user_id":"2","user_info":{"id":"2","name":"James"}},
   ...
]
```

The `channelMembers` function in `RedisClient` should handle the `hgetall` like:

```
return $this->publishClient->__call('hgetall', ["$appId:$channel"])
    ->then(function ($members) {
        // The data is expected as objects, so we need to JSON decode
        $result = [];
        foreach (array_chunk($members, 2) as $member) {
            $result[$member[0]] = json_decode($member[1], false);
        }
        return $result;
    });
```

Reply:
Oh, interesting. Good investigating! I might have missed that detail in my testing and my reading of the Redis API.

I don't think those keys are particularly useful here, do you agree that we could just drop them and just return the values? I.e. just do `$result[] = json_decode($member[1]);`

Reply:
I'd be happy to contribute an adjustment. I agree with your suggestion of ignoring the keys since the channel IDs aren't needed as keys